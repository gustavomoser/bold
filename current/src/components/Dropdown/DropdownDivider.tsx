import React from 'react'

import { Theme, useStyles } from '../../styles'

export interface DropdownDividerProps extends React.HtmlHTMLAttributes<HTMLLIElement> {}

export function DropdownDivider(props: DropdownDividerProps) {
  const { classes } = useStyles(styles)

  return <li className={classes.divider} role='separator' {...props} />
}

DropdownDivider.defaultProps = {} as Partial<DropdownDividerProps>

const styles = (theme: Theme) => ({
  divider: {
    borderBottom: `1px solid ${theme.pallete.divider}`,
    marginBottom: '0.5rem',
  },
})
