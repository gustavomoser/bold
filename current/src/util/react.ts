import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
