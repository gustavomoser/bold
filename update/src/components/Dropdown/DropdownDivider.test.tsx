import { render } from '@testing-library/react'
import React from 'react'

import { DropdownDivider } from './DropdownDivider'

it('should render correctly', () => {
  const { container } = render(<DropdownDivider />)
  expect(container).toMatchSnapshot()
})
