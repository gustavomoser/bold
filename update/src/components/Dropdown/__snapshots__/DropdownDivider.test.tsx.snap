// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-0 {
  border-bottom: 1px solid #D3D4DD;
  margin-bottom: 0.5rem;
}

<div>
  <li
    class="emotion-0"
    role="separator"
  />
</div>
`;
